sysl2/Grammar1_test.go

```diff
@@ -0,0 +1,253 @@
+package main
+
+import (
+    "testing"
+
+    sysl "anz-bank/sysl/sysl2/proto"
+)
+
+func testParser(g *sysl.Grammar, numTerms int, tokens []int, text string, expected bool, t *testing.T) {
+    p := makeParser(g, text)
+
+    if len(p.terminals) != numTerms {
+        t.Error("got incorrect number of terms", p.terminals)
+    }
+
+    for i, expected := range tokens {
+        tok := p.l.nextToken()
+        if tok != expected {
+            t.Errorf("got the wrong token at %d expected: %d, got %d", i, expected, tok)
+        }
+    }
+
+    result := p.parse(tokens[:len(tokens)-1])
+
+    if result != expected {
+        t.Error("failed to parse " + text)
+    }
+}
+
+// S –> bab | bA
+// A –> d | cA
+func TestGrammar(t *testing.T) {
+    tokens := []int{2, 3, 2, -1}
+    text := "bab"
+
+    testParser(makeGrammar1(), 4, tokens, text, true, t)
+}
+
+func TestGrammar1(t *testing.T) {
+    tokens := []int{2, 1, 1, 1, 0, -1}
+    text := "bcccd"
+
+    testParser(makeGrammar1(), 4, tokens, text, true, t)
+}
+
+func TestGrammarIncompleteInput(t *testing.T) {
+    tokens := []int{2, 3, -1}
+    text := "ba"
+
+    testParser(makeGrammar1(), 4, tokens, text, false, t)
+}
+
+func TestEXPR1(t *testing.T) {
+    text := "1 + 3 * 7"
+    tokens := []int{6, 0, 6, 2, 6, -1}
+
+    testParser(makeEXPR(), 7, tokens, text, true, t)
+}
+
+func TestOBJ(t *testing.T) {
+    text := "{}"
+    tokens := []int{0, 2, -1}
+
+    testParser(makeRepeatSeq(makeQuantifierZeroPlus()), 4, tokens, text, true, t)
```

Comment:
I understand that `makeRepeatSeq ` creates a grammar for {} | { NUM (, NUM)*}, but I didn't manage to work out what `makeQuantifierZeroPlus` actually does.